resteasy-jaxrs/src/main/java/org/jboss/resteasy/core/StringParameterInjector.java

```diff
@@ -80,40 +80,39 @@ protected void initialize(Class type, Type genericType, String paramName, Class
       baseType = type;
       baseGenericType = genericType;
 
-      if (type.isArray()) baseType = type.getComponentType();
-      if (List.class.isAssignableFrom(type))
-      {
-         isCollection = true;
-         collectionType = ArrayList.class;
-      }
-      else if (SortedSet.class.isAssignableFrom(type))
-      {
-         isCollection = true;
-         collectionType = TreeSet.class;
-      }
-      else if (Set.class.isAssignableFrom(type))
+      if (type.isArray())
       {
-         isCollection = true;
-         collectionType = HashSet.class;
+    	  baseType = type.getComponentType();
       }
-      if (isCollection)
+     
+      if (!baseType.isPrimitive())
       {
-         if (genericType != null && genericType instanceof ParameterizedType)
+         paramConverter = factory.getParamConverter(baseType, baseGenericType, annotations);
+         if (paramConverter != null) 
          {
-            ParameterizedType zType = (ParameterizedType) genericType;
-            baseType = Types.getRawType(zType.getActualTypeArguments()[0]);
-            baseGenericType = zType.getActualTypeArguments()[0];
+        	 return;
          }
-         else
-         {
-            baseType = String.class;
-            baseGenericType = null;
+         
+         collectionType = convertParameterTypeToCollectionType();
+         if(collectionType != null){
+        	 isCollection = true;
+	         if (genericType != null && genericType instanceof ParameterizedType)
+	         {
+	            ParameterizedType zType = (ParameterizedType) baseGenericType;
+	            baseType = Types.getRawType(zType.getActualTypeArguments()[0]);
+	            baseGenericType = zType.getActualTypeArguments()[0];
+	         }
+	         else
+	         {
+	            baseType = String.class;
+	            baseGenericType = null;
+	         }
+	         paramConverter = factory.getParamConverter(baseType, baseGenericType, annotations);
+	         if (paramConverter != null) 
+	         {
+	        	 return;
+	         }
          }
-      }
-      if (!baseType.isPrimitive())
-      {
-         paramConverter = factory.getParamConverter(baseType, baseGenericType, annotations);
-         if (paramConverter != null) return;
 
          unmarshaller = factory.createStringParameterUnmarshaller(baseType);

```

Comment:
Which version of the JAX-RS spec are you looking at? Version 2.0 Final Release says

1. Types for which a ParamConverter is available via a registered ParamConverterProvider . See
Javadoc for these classes for more information.
2. Primitive types.
3. Types that have a constructor that accepts a single String argument.
4. Types that have a static method named valueOf or fromString with a single String argument
that return an instance of the type. If both methods are present then valueOf MUST be used unless
the type is an enum in which case fromString MUST be used 1 .
5. List<T> , Set<T> , or SortedSet<T> , where T satisfies 3 or 4 above.